### PR TITLE
SWATCH-969: Replace ProductId enum with class that checks against product-config

### PR DESCRIPTION
**Pull request by kahowell**
Jira issue: [SWATCH-969](https://issues.redhat.com/browse/SWATCH-969)

Description
===========

Remove productId enum in favor of a class that validates against product-config (extremely similar
to #2464). I made these changes on top of that PR in order to avoid merge conflicts, so #2464 must
land first.

Testing
=======

Setup
-----

1. Deploy the service:

```shell
DEV_MODE=true \
  ./gradlew :bootRun
```

2. Ensure opted in:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/opt-in \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Verification - Happy Path
-------------------------

Try the following APIs, expecting successful responses...

capacity (deprecated):

```shell
http ':8000/api/rhsm-subscriptions/v1/capacity/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

capacity:

```shell
http ':8000/api/rhsm-subscriptions/v1/capacity/products/RHEL for x86/Sockets' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

hosts:

```shell
http ':8000/api/rhsm-subscriptions/v1/hosts/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

instances:

```shell
http ':8000/api/rhsm-subscriptions/v1/instances/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

subscriptions:

```shell
http ':8000/api/rhsm-subscriptions/v1/subscriptions/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally (deprecated):

```shell
http ':8000/api/rhsm-subscriptions/v1/tally/products/RHEL for x86' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally:

```shell
http ':8000/api/rhsm-subscriptions/v1/tally/products/RHEL for x86/Sockets' \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Verification - Bad Requests
---------------------------

Try the following APIs, expected 404 w/ an error message explaining the missing productId...

capacity (deprecated):

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

capacity:

```shell
http :8000/api/rhsm-subscriptions/v1/capacity/products/NotAProduct/Cores \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

hosts:

```shell
http :8000/api/rhsm-subscriptions/v1/hosts/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

instances:

```shell
http :8000/api/rhsm-subscriptions/v1/instances/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

subscriptions:

```shell
http :8000/api/rhsm-subscriptions/v1/subscriptions/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally (deprecated):

```shell
http :8000/api/rhsm-subscriptions/v1/tally/products/NotAProduct \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

tally:

```shell
http :8000/api/rhsm-subscriptions/v1/tally/products/NotAProduct/Cores \
  granularity==Daily \
  beginning==2023-07-21T17:32:28Z \
  ending==2023-07-22T17:32:28Z \
  x-rh-identity:$(echo '{"identity":{"type":"User","internal":{"org_id":"org123"}}}' | base64 -w0)
```

Draft until #2464 lands.

### CODE REVIEW

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/ProductId.java
Comment (awood) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/ProductId.java:56]: Doesn't `@Data` include a `toString` as well?  I imagine ours would take precedence over a generated method, but a comment here noting that fact (and why we want a custom implementation) would be useful.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java:116]: I'd like to do these checks with a custom validator.  It reduces the boilerplate and it should be pretty easy to write.  We have several examples already in the `swatch-core/src/main/java/org/candlepin/subscriptions/validator` package.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java:192]: And we might as well do a `MetricId` validator while we're at it.

**Review by awood — CHANGES_REQUESTED** [review]
Let me know if you need any help with writing a validator.  I've done a handful before.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java:116]: I understand your request, and I looked into doing this just now, and there's one small reason I'm nacking this request: Validators don't let us keep the looked-value, so we'd end up effectively scanning the config twice to do essentially the same validation. I wish that JSR-303 had some sort of generic "Converter" interface for this. It's tempting to implement a JAX-RS converter, but that doesn't fit with out OpenAPI generation specs.

File: src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/resource/CapacityResource.java:192]: Nacking for above reason

**Comment by ntkathole** [comment]
Feel free to merge this without QE ack once CI tests passed. Card is marked for QE to update api spec from automation, which can be done after merge. @Aurobinda55 will take a look once back from PTO

**Comment by kahowell** [comment]
Following @awood's suggestion, we figured out how to centralize the conversion/error handling in a more JAX-RS-native way. As a result, the error code changed to 404, which is actually more RESTful.

@awood please take another look when you get a chance.

File: swatch-common-resteasy/src/main/java/com/redhat/swatch/common/resteasy/MetricIdParamConverterProvider.java
Comment (awood) [inline swatch-common-resteasy/src/main/java/com/redhat/swatch/common/resteasy/MetricIdParamConverterProvider.java:46]: Please correct this comment.

**Review by awood — APPROVED** [review]

**Comment by awood** [comment]
> @awood please take another look when you get a chance.

There's one copy-paste error in a comment, but otherwise things look fine.  If you can fix it before you hit merge, great.  If not, I'm not going to sweat it.

**Comment by kahowell** [comment]
1da9563 to 8f76bad: spotlessApply

**Pull request merged**
